feat(ecmascript): complete may_have_side_effects

sapphi-red opened the pull request:
Ported tests from closure compiler and reviewed the behavior of may_have_side_effects and added additional tests.

sapphi-red commented on crates/oxc_ecmascript/src/constant_evaluation/mod.rs:
```diff
@@ -327,6 +327,10 @@ pub trait ConstantEvaluation<'a>: MayHaveSideEffects {
                     if matches!(name, "Object" | "Number" | "Boolean" | "String")
                         && self.is_global_reference(right_ident)
                     {
+                        let left_ty = ValueType::from(left);
+                        if left_ty.is_undetermined() {
+                            return None;
+                        }
```
Without this change, this test failed because `var x; foo(!1)` was output.
https://github.com/oxc-project/oxc/blob/d6d13dd296fd365d4229e02d72a6658266b914c7/crates/oxc_minifier/src/peephole/fold_constants.rs#L1724

sapphi-red commented on crates/oxc_minifier/README.md:
```diff
@@ -29,6 +29,8 @@ The compressor is responsible for rewriting statements and expressions for minim
   - Examples that breaks this assumption: `(() => { console.log(v); let v; })()`
 - `with` statement is not used
   - Examples that breaks this assumption: `with (Math) { console.log(PI); }`
+- `.toString()`, `.valueOf()`, `[Symbol.toPrimitive]()` are side-effect free
+  - Examples that breaks this assumption: `{ toString() { console.log('sideeffect') } }`
```
uglify, terser, closure compiler makes this assumption. All of them does not say anything about `Symbol.toPrimitive`, but it doesn't make sense without it and I assume they do so.

sapphi-red commented on crates/oxc_ecmascript/src/side_effects/may_have_side_effects.rs:
```diff
@@ -108,4 +144,529 @@ pub trait MayHaveSideEffects {
+        test("a + b", false);
+        test("a() + b", true);
+
+        // b maybe not a object
+        // b maybe a proxy that has a side effectful "has" trap
+        test("a in b", true);
+        // b maybe not a function
+        // b[Symbol.hasInstance] may have a side effect
+        // a maybe a proxy that has a side effectful "getPrototypeOf" trap
+        test("a instanceof b", true);
```
esbuild preserves these sideeffects ([try](https://esbuild.github.io/try/#dAAwLjI0LjIALS1taW5pZnkgLS10cmVlLXNoYWtpbmcAZXhwb3J0IGZ1bmN0aW9uIGluc3RhbmNlb2ZPcGVyYXRpb24oKSB7CiAgY29uc3QgdjEgPSB7IFtTeW1ib2wuaGFzSW5zdGFuY2VdKCkgeyBjb25zb2xlLmxvZygic2lkZWVmZmVjdCIpIH0gfTsKICBjb25zdCB2MiA9IG5ldyBQcm94eSh7fSwgeyBnZXRQcm90b3R5cGVPZigpIHsgY29uc29sZS5sb2coInNpZGVlZmZlY3QiKSB9IH0pOwogIDEgaW5zdGFuY2VvZiB2MTsKICAxIGluc3RhbmNlb2YgdjI7CiAgMSBpbnN0YW5jZW9mIHsgW1N5bWJvbC5oYXNJbnN0YW5jZV0oKSB7IGNvbnNvbGUubG9nKCJzaWRlZWZmZWN0IikgfSB9OwogIG5ldyBQcm94eSh7fSwgeyBnZXRQcm90b3R5cGVPZigpIHsgY29uc29sZS5sb2coInNpZGVlZmZlY3QiKTsgcmV0dXJuIG51bGwgfSB9KSBpbnN0YW5jZW9mIE9iamVjdDsKfQoKZXhwb3J0IGZ1bmN0aW9uIGluT3BlcmF0aW9uKCkgewogIGNvbnN0IHYgPSBuZXcgUHJveHkoe30sIHsgaGFzKCkgeyBjb25zb2xlLmxvZygic2lkZWVmZmVjdCIpIH0gfSk7CiAgMSBpbiB2OwogIDEgaW4gbmV3IFByb3h5KHt9LCB7IGhhcygpIHsgY29uc29sZS5sb2coInNpZGVlZmZlY3QiKSB9IH0pOwp9Cg)).
SWC and terser doesn't ([SWC playground](https://play.swc.rs/?version=1.10.12&code=H4sIAAAAAAAAA6VRwQqCQBC9C%2F7D4EkhBL1KH9DJoGN0UJs1Y9uRdTVF%2FPdGCRI0orrNvH3z9r0ZbEvSBkStMlOQgkJVJlEZkohL1MmIuR70tgWQEb9BE8AWejgeultK0r8k1e45chqJE4sk%2BpJy16mKM6IQmBnHgwGGaKYTso7CO%2Bw1tZ3bDxseztFwa8h0Jcbis543CQYz1%2BxvBQuX2F8R%2FvEdgUZTawWqlnIKMbcVp1dm8SeDbdkWLq7z7iory%2BRc32wQmlf5sxS7fgBhmzALUgIAAA%3D%3D&config=H4sIAAAAAAAAA41VwW7bMAy99ysCn3doeyiGfcBu%2BwZBsShHmSwaIuUmKPLvoxU7TRva2CWI%2BfRIkXykPp52u%2BZIbfNr9yF%2F5WOwmSDfvsVC58T2JJYG2t5Sm8PAzY8FPdIEeRsJqulyRRq2uQOuLHp9fn2ZGU1EJFgYs60PKfjzfcwW%2ByED0Z1NrOKy9JCYvvJnLOP7BHAu9%2FY9YgSbNhBjyYTE0EHWHLcYox0IzGiz4mW6qc2BUAsxgYXBmSHjoOLJBQ6YJOYj6sA606IDBQoZWg4jaDSJJbREkp6ST4Ud7EvX1T5%2FY8NoY7GsxIRTbYncVvF6wEBsfElaCa%2FgSg2u4Fzc78zgTQYuOT3yjhjSSk%2F%2BAkgFoiVKtgfNbz3hRU9rbL%2FJDMmLZPms4KJvLcsEnRTVhOCVyk6VgcxB62YGV1qYKttq15nhlfJRcGDAe9GK4preA7cHLSifB0CvANJf6zVVXQFzm8IVfBqIDfi3ZMm6wOYTveXDOkrnfo9xI0APfEC3cUBawbgOZ9kSp2EdL8mBSAOceqRQBR6XgAwAo4l1Xz5oQ8ZDPJou4v5zTcwHLrc93NvU1Xm%2F25eMQ4QR4pqM%2F2NENlG52zipepnsh%2BmFn5oorQxeeHn%2B8mxIKk%2FLb02q6dGVu4SqJq%2FPyVvzeWh5OZaqNYH%2BLMRap8s%2FdrgRoOMGAAA%3D)).

sapphi-red commented on crates/oxc_ecmascript/src/side_effects/may_have_side_effects.rs:
```diff
@@ -108,4 +144,529 @@ pub trait MayHaveSideEffects {
+
+    #[test]
+    fn tests() {
+        // This actually have a side effect, but this treated as side-effect free.
+        test("'' + { toString() { console.log('sideeffect') } }", false);
+        test("'' + { valueOf() { console.log('sideeffect') } }", false);
+        test("'' + { [s]() { console.log('sideeffect') } }", false); // assuming s is Symbol.toPrimitive
```
esbuild preserves these sideeffects ([try](https://esbuild.github.io/try/#dAAwLjI0LjIALS1taW5pZnkgLS10cmVlLXNoYWtpbmcAZXhwb3J0IGZ1bmN0aW9uIHN0cmluZ0FkZCgpIHsKICBjb25zdCB2MSA9IHsgdG9TdHJpbmcoKSB7IGNvbnNvbGUubG9nKCJzaWRlZWZmZWN0IikgfSB9OwogIGNvbnN0IHYyID0geyB2YWx1ZU9mKCkgeyBjb25zb2xlLmxvZygic2lkZWVmZmVjdCIpIH0gfTsKICBjb25zdCB2MyA9IHsgW1N5bWJvbC50b1ByaW1pdGl2ZV0oKSB7IGNvbnNvbGUubG9nKCJzaWRlZWZmZWN0IikgfSB9OwogICcnK3YxOwogICcnK3YyOwogICcnK3YzOwoKICAnJyt7IHRvU3RyaW5nKCkgeyBjb25zb2xlLmxvZygic2lkZWVmZmVjdCIpIH0gfTsKICAnJyt7IHZhbHVlT2YoKSB7IGNvbnNvbGUubG9nKCJzaWRlZWZmZWN0IikgfSB9OwogICcnK3sgW1N5bWJvbC50b1ByaW1pdGl2ZV0oKSB7IGNvbnNvbGUubG9nKCJzaWRlZWZmZWN0IikgfSB9Owp9)).
SWC and terser doesn't ([SWC playground](https://play.swc.rs/?version=1.10.12&code=H4sIAAAAAAAAA52PsQrCMBRFd8F%2FeHRpi1BoO4qDX6DQURw0eSmBNE%2Ba16CU%2FLsapDgatwOHc%2BHi%2FUYjg5qsYE0WHI%2Fa9nspixLm9QpAkHUMvoYdzMDURf%2BW0ZDBylBfZE5LRKVQcFZCgLD9apvY%2BouZ8KDS0jamp%2B4xXMlUTMdRD5q1x%2FOPO3m%2B8fVCzULtiz6cfCs2aXdi8veN8AQ0ZGnspgEAAA%3D%3D&config=H4sIAAAAAAAAA41VwW7bMAy99ysCn3doeyiGfcBu%2BwZBsShHmSwaIuUmKPLvoxU7TRva2CWI%2BfRIkXykPp52u%2BZIbfNr9yF%2F5WOwmSDfvsVC58T2JJYG2t5Sm8PAzY8FPdIEeRsJqulyRRq2uQOuLHp9fn2ZGU1EJFgYs60PKfjzfcwW%2ByED0Z1NrOKy9JCYvvJnLOP7BHAu9%2FY9YgSbNhBjyYTE0EHWHLcYox0IzGiz4mW6qc2BUAsxgYXBmSHjoOLJBQ6YJOYj6sA606IDBQoZWg4jaDSJJbREkp6ST4Ud7EvX1T5%2FY8NoY7GsxIRTbYncVvF6wEBsfElaCa%2FgSg2u4Fzc78zgTQYuOT3yjhjSSk%2F%2BAkgFoiVKtgfNbz3hRU9rbL%2FJDMmLZPms4KJvLcsEnRTVhOCVyk6VgcxB62YGV1qYKttq15nhlfJRcGDAe9GK4preA7cHLSifB0CvANJf6zVVXQFzm8IVfBqIDfi3ZMm6wOYTveXDOkrnfo9xI0APfEC3cUBawbgOZ9kSp2EdL8mBSAOceqRQBR6XgAwAo4l1Xz5oQ8ZDPJou4v5zTcwHLrc93NvU1Xm%2F25eMQ4QR4pqM%2F2NENlG52zipepnsh%2BmFn5oorQxeeHn%2B8mxIKk%2FLb02q6dGVu4SqJq%2FPyVvzeWh5OZaqNYH%2BLMRap8s%2FdrgRoOMGAAA%3D)) (intended behavior).

sapphi-red commented on crates/oxc_ecmascript/src/side_effects/may_have_side_effects.rs:
```diff
@@ -108,4 +144,529 @@ pub trait MayHaveSideEffects {
+        test("'' + { [s]() { console.log('sideeffect') } }", false); // assuming s is Symbol.toPrimitive
+
+        // FIXME: actually these have a side effect, but it's ignored now
+        test("+s", false); // assuming s is a Symbol
+        test("+b", false); // assuming b is a BitInt
+        test("+{ valueOf() { return Symbol() } }", false);
+        test("~s", false); // assuming s is a Symbol
```
esbuild preserves these sideeffects ([try](https://esbuild.github.io/try/#dAAwLjI0LjIALS1taW5pZnkgLS10cmVlLXNoYWtpbmcAZXhwb3J0IGZ1bmN0aW9uIHVuYXJ5KCkgewogIGNvbnN0IHMgPSBTeW1ib2woKTsKICBjb25zdCB2ID0geyB2YWx1ZU9mKCkgeyByZXR1cm4gU3ltYm9sKCkgfSB9OwogIGNvbnN0IGIgPSAwbjsKICArczsKICArdjsKICArYjsKICB+czsKCiAgK1N5bWJvbCgpOwogICt7IHZhbHVlT2YoKSB7IHJldHVybiBTeW1ib2woKSB9IH07CiAgKzBuOwogIH5TeW1ib2woKTsKfQ)).
SWC and terser doesn't ([SWC playground](https://play.swc.rs/?version=1.10.12&code=H4sIAAAAAAAAA0utKMgvKlFIK81LLsnMz1MozUssqtTQVKjm5VJQSM7PKy5RKFawVQiuzE3Kz9HQtEYIlwGFqxXKEnNKU%2F3TQDoUilJLSovy4GoVahVqkdQnAdUb5IEFtIshVBmESgJTdSBBMB%2FFMm3i7NCGGl2HpLkWAN4PpEXdAAAA&config=H4sIAAAAAAAAA41VwW7bMAy99ysCn3doeyiGfcBu%2BwZBsShHmSwaIuUmKPLvoxU7TRva2CWI%2BfRIkXykPp52u%2BZIbfNr9yF%2F5WOwmSDfvsVC58T2JJYG2t5Sm8PAzY8FPdIEeRsJqulyRRq2uQOuLHp9fn2ZGU1EJFgYs60PKfjzfcwW%2ByED0Z1NrOKy9JCYvvJnLOP7BHAu9%2FY9YgSbNhBjyYTE0EHWHLcYox0IzGiz4mW6qc2BUAsxgYXBmSHjoOLJBQ6YJOYj6sA606IDBQoZWg4jaDSJJbREkp6ST4Ud7EvX1T5%2FY8NoY7GsxIRTbYncVvF6wEBsfElaCa%2FgSg2u4Fzc78zgTQYuOT3yjhjSSk%2F%2BAkgFoiVKtgfNbz3hRU9rbL%2FJDMmLZPms4KJvLcsEnRTVhOCVyk6VgcxB62YGV1qYKttq15nhlfJRcGDAe9GK4preA7cHLSifB0CvANJf6zVVXQFzm8IVfBqIDfi3ZMm6wOYTveXDOkrnfo9xI0APfEC3cUBawbgOZ9kSp2EdL8mBSAOceqRQBR6XgAwAo4l1Xz5oQ8ZDPJou4v5zTcwHLrc93NvU1Xm%2F25eMQ4QR4pqM%2F2NENlG52zipepnsh%2BmFn5oorQxeeHn%2B8mxIKk%2FLb02q6dGVu4SqJq%2FPyVvzeWh5OZaqNYH%2BLMRap8s%2FdrgRoOMGAAA%3D)).

sapphi-red commented on crates/oxc_ecmascript/src/side_effects/may_have_side_effects.rs:
```diff
@@ -108,4 +144,529 @@ pub trait MayHaveSideEffects {
+        // FIXME: actually these have a side effect, but it's ignored now
+        // same for -, *, /, %, **, <<, >>
+        test("'' + s", false); // assuming s is Symbol
+        test("0 + b", false); // assuming b is a BitInt
```
esbuild preserves this sideeffect ([try](https://esbuild.github.io/try/#dAAwLjI0LjIALS1taW5pZnkgLS10cmVlLXNoYWtpbmcAZXhwb3J0IGZ1bmN0aW9uIG51bWJlckFkZCgpIHsKICBjb25zdCBiID0gMG47CiAgMCArIGI7CiAgMCArIDBuOwp9)).
SWC and terser doesn't ([SWC playground](https://play.swc.rs/?version=1.10.12&code=H4sIAAAAAAAAA0utKMgvKlFIK81LLsnMz1PIK81NSi1yTEnR0FSo5uVSUEjOzysuUUhSsFUwyLMGCRgoaCskwVlgwVoAenVoaEYAAAA%3D&config=H4sIAAAAAAAAA41VwW7bMAy99ysCn3doeyiGfcBu%2BwZBsShHmSwaIuUmKPLvoxU7TRva2CWI%2BfRIkXykPp52u%2BZIbfNr9yF%2F5WOwmSDfvsVC58T2JJYG2t5Sm8PAzY8FPdIEeRsJqulyRRq2uQOuLHp9fn2ZGU1EJFgYs60PKfjzfcwW%2ByED0Z1NrOKy9JCYvvJnLOP7BHAu9%2FY9YgSbNhBjyYTE0EHWHLcYox0IzGiz4mW6qc2BUAsxgYXBmSHjoOLJBQ6YJOYj6sA606IDBQoZWg4jaDSJJbREkp6ST4Ud7EvX1T5%2FY8NoY7GsxIRTbYncVvF6wEBsfElaCa%2FgSg2u4Fzc78zgTQYuOT3yjhjSSk%2F%2BAkgFoiVKtgfNbz3hRU9rbL%2FJDMmLZPms4KJvLcsEnRTVhOCVyk6VgcxB62YGV1qYKttq15nhlfJRcGDAe9GK4preA7cHLSifB0CvANJf6zVVXQFzm8IVfBqIDfi3ZMm6wOYTveXDOkrnfo9xI0APfEC3cUBawbgOZ9kSp2EdL8mBSAOceqRQBR6XgAwAo4l1Xz5oQ8ZDPJou4v5zTcwHLrc93NvU1Xm%2F25eMQ4QR4pqM%2F2NENlG52zipepnsh%2BmFn5oorQxeeHn%2B8mxIKk%2FLb02q6dGVu4SqJq%2FPyVvzeWh5OZaqNYH%2BLMRap8s%2FdrgRoOMGAAA%3D)).

sapphi-red commented on crates/oxc_ecmascript/src/side_effects/may_have_side_effects.rs:
```diff
@@ -108,4 +144,529 @@ pub trait MayHaveSideEffects {
+        test("0 + b", false); // assuming b is a BitInt
+
+        // FIXME: actually these throws an error, but it's ignored now
+        test("1n ** (-1n)", false);
+        test("1n / 0n", false);
+        test("1n % 0n", false);
+        test("0n >>> 1n", false); // >>> throws an error even when both operands are bigint
```
esbuild preserves these sideeffects ([try](https://esbuild.github.io/try/#dAAwLjI0LjIALS1taW5pZnkgLS10cmVlLXNoYWtpbmcAZXhwb3J0IGZ1bmN0aW9uIGJpZ2ludE9wZXJhdGlvbnMoKSB7CiAgMW4gKiogKC0xbik7CiAgMW4gLyAwbjsKICAxbiAlIDBuOwogIDBuID4+PiAxbjsKfQ)).
SWC doesn't ([SWC playground](https://play.swc.rs/?version=1.10.12&code=H4sIAAAAAAAAA0utKMgvKlFIK81LLsnMz1NIykzPzCvxL0gtSgTxizU0Fap5uRQUDPMUtLQUNHQN8zStoXx9BYM8GFsVxjbIU7CzswMKAXm1AEi9g6ddAAAA&config=H4sIAAAAAAAAA41VwW7bMAy99ysCn3doeyiGfcBu%2BwZBsShHmSwaIuUmKPLvoxU7TRva2CWI%2BfRIkXykPp52u%2BZIbfNr9yF%2F5WOwmSDfvsVC58T2JJYG2t5Sm8PAzY8FPdIEeRsJqulyRRq2uQOuLHp9fn2ZGU1EJFgYs60PKfjzfcwW%2ByED0Z1NrOKy9JCYvvJnLOP7BHAu9%2FY9YgSbNhBjyYTE0EHWHLcYox0IzGiz4mW6qc2BUAsxgYXBmSHjoOLJBQ6YJOYj6sA606IDBQoZWg4jaDSJJbREkp6ST4Ud7EvX1T5%2FY8NoY7GsxIRTbYncVvF6wEBsfElaCa%2FgSg2u4Fzc78zgTQYuOT3yjhjSSk%2F%2BAkgFoiVKtgfNbz3hRU9rbL%2FJDMmLZPms4KJvLcsEnRTVhOCVyk6VgcxB62YGV1qYKttq15nhlfJRcGDAe9GK4preA7cHLSifB0CvANJf6zVVXQFzm8IVfBqIDfi3ZMm6wOYTveXDOkrnfo9xI0APfEC3cUBawbgOZ9kSp2EdL8mBSAOceqRQBR6XgAwAo4l1Xz5oQ8ZDPJou4v5zTcwHLrc93NvU1Xm%2F25eMQ4QR4pqM%2F2NENlG52zipepnsh%2BmFn5oorQxeeHn%2B8mxIKk%2FLb02q6dGVu4SqJq%2FPyVvzeWh5OZaqNYH%2BLMRap8s%2FdrgRoOMGAAA%3D)).
terser errors on minification (`undefined must be positive`, `Division by zero`).

Boshen commented on crates/oxc_ecmascript/Cargo.toml:
```diff
@@ -28,6 +28,10 @@ oxc_syntax = { workspace = true, features = ["to_js_string"] }
 
+[dev-dependencies]
+oxc_allocator = { workspace = true }
+oxc_parser = { workspace = true }
```
This creates a cyclic dependency with `oxc_parser` (an unfortunate cargo limitation). This is why all the tests are in `oxc_minifier` :-(

Boshen replied:
The cyclic dependency breaks release, but nothing happens during development :thrug:

sapphi-red replied:
Ah, I see. I was wondering why this wasn't done as it worked. 😅
Let me move the tests.

Boshen reviewed (commented):
Sorry for the request changes :-(

> This creates a cyclic dependency with oxc_parser (an unfortunate cargo limitation). This is why all the tests are in oxc_minifier :-(

Boshen reviewed (changes requested):
Sorry for the request changes :-(

> This creates a cyclic dependency with oxc_parser (an unfortunate cargo limitation). This is why all the tests are in oxc_minifier :-(

Boshen commented:
### Merge activity

* **Feb 4, 10:49 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

sapphi-red committed:
feat(ecmascript): complete may_have_side_effects (#8855)

Ported tests from closure compiler and reviewed the behavior of may_have_side_effects and added additional tests.
